core/market/decentralized/src/db/models/events.rs

```diff
@@ -0,0 +1,123 @@
+use chrono::{DateTime, NaiveDateTime, Utc};
+use diesel::backend::Backend;
+use diesel::deserialize;
+use diesel::serialize::Output;
+use diesel::sql_types::Integer;
+use diesel::types::{FromSql, ToSql};
+use num_derive::FromPrimitive;
+use num_traits::FromPrimitive;
+use thiserror::Error;
+
+use ya_client::model::market::event::RequestorEvent;
+use ya_client::model::ErrorMessage;
+use ya_persistence::executor::{DbExecutor, Error as DbError};
+
+use super::SubscriptionId;
+use crate::db::dao::ProposalDao;
+use crate::db::models::{OwnerType, Proposal};
+use crate::db::schema::market_requestor_event;
+
+#[derive(Error, Debug)]
+pub enum EventError {
+    #[error("Proposal [{0}] not found.")]
+    ProposalNotFound(String),
+    #[error("Failed get proposal from database. Error: {0}.")]
+    FailedGetProposal(DbError),
+    #[error("Unexpected error: {0}.")]
+    InternalError(#[from] ErrorMessage),
+}
+
+#[derive(FromPrimitive, AsExpression, FromSqlRow, PartialEq, Debug, Clone, Copy)]
+#[sql_type = "Integer"]
+pub enum EventType {
+    ProviderProposal = 1001,
+    ProviderAgreement = 1002,
+    ProviderPropertyQuery = 1003,
+    RequestorProposal = 2001,
+    RequestorPropertyQuery = 2002,
+}
+
+/// TODO: We need two separate tables for Provider and Requestor events.
+///  This way we can avoid storing additional field with flag.
+#[derive(Clone, Debug, Queryable)]
+pub struct MarketEvent {
+    pub id: i32,
+    pub subscription_id: SubscriptionId,
+    pub timestamp: NaiveDateTime,
+    pub event_type: EventType,
+    /// It can be Proposal, Agreement or structure,
+    /// that will represent PropertyQuery.
+    pub artifact_id: String,
+}
+
+#[derive(Clone, Debug, Insertable)]
+#[table_name = "market_provider_event"]
+#[table_name = "market_requestor_event"]
+pub struct NewMarketEvent {
+    pub subscription_id: SubscriptionId,
+    pub event_type: EventType,
+    pub artifact_id: String,
+}
+
+impl MarketEvent {
+    pub fn from_proposal(proposal: &Proposal, role: OwnerType) -> NewMarketEvent {
+        NewMarketEvent {
+            subscription_id: proposal.negotiation.subscription_id.clone(),
+            event_type: match role {
```

Comment:
again, here is matching. So we did not avoid matching of roles even though tables are denormalized.

Reply:
We avoided many things. When you use already created types, you don't have to do any matching. This is only creation function. I could split it into 2 separate functions with different names and you would have any match, but i didn't want to duplicate code.